library/alloc/src/string.rs

```diff
@@ -913,10 +913,9 @@ impl String {
     /// This is highly unsafe, due to the number of invariants that aren't
     /// checked:
     ///
-    /// * The memory at `buf` needs to have been previously allocated by the
-    ///   same allocator the standard library uses, with a required alignment of exactly 1.
+    /// * unless `capacity` is 0, `buf` must have been allocated using the global allocator with an alignment of 1 and a capacity of `capacity`.
```

Comment:
So do we really want to promise that *any* non-null pointer is okay if capacity is 0?

Also, "allocated using the global allocator with [...] a capacity" is odd -- the allocation ABI doesn't talk about "capacity", it talks about "size". So this should probably say "and a size of `capacity`".

Reply:
Yeah, I kinda think we don't, and should specify the `NonNull::dangling()` pointer if we accept any.

Reply:
Regarding the point about using "size" rather than "capacity"; note that this is consistent with some of the current language on `Vec::from_raw_parts` "`capacity` needs to be the capacity that the pointer was allocated with". Maybe that should be changed as well then?

Regarding the requirement on `String::from_raw_parts`; what about doing the same as in `Vec::from_raw_parts`, making the alignment requirement always present? That would allow making your own dangling pointers, but I don't see the motivation for restricting the creation of dangling pointers to a specific API here.

Reply:
> Regarding the requirement on `String::from_raw_parts`; what about doing the same as in `Vec::from_raw_parts`, making the alignment requirement always present? That would allow making your own dangling pointers, but I don't see the motivation for restricting the creation of dangling pointers to a specific API here.

So that we can change the value that we're expecting as the dangling pointer.

Reply:
> So that we can change the value that we're expecting as the dangling pointer.

`NonNull::dangling` does specifically specify that it should never be used as a sentinel value, as it may also be a valid pointer.

> what about doing the same as in Vec::from_raw_parts, making the alignment requirement always present?

is it possible to have a pointer that isn't byte aligned?  I considered saying "the pointer must have an alignment of 1", but isn't that a given? maybe it can change the behavior of allocators?  but, that shouldn't matter for a dangling pointer that is never freed.

Reply:
> `NonNull::dangling` does specifically specify that it should never be used as a sentinel value, as it may also be a valid pointer.

See prior comments about the standard library being allowed to write unfair contracts and rely on its own internal implementation details in ways you cannot.

Reply:
I may be being a bit silly here, but it just still doesn't seem like a good idea to bless all pointers thusly.

Reply:
> See prior comments about the standard library being allowed to write unfair contracts and rely on its own internal implementation details in ways you cannot.

except that the behavior of allocators is not an internal implementation detail of the standard library.  even if it was, `NonNull::dangling` can be manipulated into taking any value by changing the size of `T`.  therefore, even with the special privilege of being the standard library, using `NonNull::dangling` as a sentinel value is still a logic error.